File: src/time/time.cpp
#include "time.h"
#include "../controller.h"
#include "../coroutine.h"

namespace flame::time
{

    static std::chrono::time_point<std::chrono::system_clock> time_system;
	static std::chrono::time_point<std::chrono::steady_clock> time_steady;

    static php::value sleep(php::parameters& params)
    {
        coroutine_handler ch {coroutine::current};
        boost::asio::steady_timer tm(gcontroller->context_x);
        int ms = static_cast<int>(params[0]);
        if(ms < 0) ms = 1;
        tm.expires_from_now(std::chrono::milliseconds(ms));
        tm.async_wait(ch);
        return nullptr;
    }


    std::chrono::time_point<std::chrono::system_clock> now()
    {
        std::chrono::milliseconds diff = std::chrono::duration_cast<std::chrono::milliseconds>(
            std::chrono::steady_clock::now() - time_steady);
        return time_system + diff;
    }
    php::string iso(const std::chrono::time_point<std::chrono::system_clock>& now) {
        php::string data(19);
        std::time_t t = std::chrono::system_clock::to_time_t(now);
        struct tm *m = std::localtime(&t);
        sprintf(data.data(), "%04d-%02d-%02d %02d:%02d:%02d",
                1900 + m->tm_year,
                1 + m->tm_mon,
                m->tm_mday,
                m->tm_hour,
                m->tm_min,
                m->tm_sec);
        return std::move(data);
    }
    php::string iso()
    {
        return iso(now());
    }

    static php::value now(php::parameters &params)
    {
        return std::chrono::duration_cast<std::chrono::milliseconds>(now().time_since_epoch()).count();
    }
    static php::value iso(php::parameters &params)
    {
        return iso();
    }

    void declare(php::extension_entry& ext)
    {
        time_steady = std::chrono::steady_clock::now();
		time_system = std::chrono::system_clock::now();

        ext
        .function<sleep>("flame\\time\\sleep",
        {
            {"duration", php::TYPE::INTEGER},
        })
        // 毫秒时间戳
        .function<now>("flame\\time\\now")
        // 标准时间 YYYY-mm-dd HH:ii:ss
        .function<iso>("flame\\time\\iso");

    }

}

File: doc/time.php
<?php
/**
 * 时间相关函数
 */
namespace flame\time;
/**
 * 获取毫秒时间戳
 * 注意：
 *  1. 此时间戳使用相对比直接获取时钟消耗更小的方式计算系统时间；
 */
function now():int {
    return 1557659996427;
}
/**
 * 获取当前时区标准时间 "YYYY-mm-dd HH:ii:ss"
 */
function iso():string {
    return "2019-03-31 21:40:25";
}
/**
 * 获取UTC标准时间 "YYYY-mm-ddTHH:ii:ssZ"
 */
function utc():string {
    return "2020-01-15T17:30:24Z";
}
/**
 * 暂停、挂起当前协程（调度执行其他活跃协程），并在 $ms 毫秒后恢复
 * @param int $ms 毫秒数，最小值 1 毫秒
 * 
 * 注意：
 *  1. 使用 PHP 内置的 `sleep()`/`usleep()` 函数会阻塞所有协程；
 */
function sleep(int $ms) {}
